lib/iris/fileformats/pp.py

```diff
@@ -1040,7 +1040,8 @@ def _data_bytes_to_shaped_array(data_bytes, lbpack, boundary_packing,
 
     # Mask the array?
     if mdi in data:
-        data = array_masked_to_nans(data, data == mdi)
+        data = np.ma.masked_array(data, data == mdi)
```

Comment:
i think in my previous implementation this was all done in line and the reassignment happened conditionally

in factoring this out to share the type change code, i think this performance benefit was lost; I think that your refactoring just highlighted its loss

I am tempted to suggest that sharing code is one module is not always the best thing to do and this may be a context where it is easier to just process the data in line and accept and recognise a bit of shared logic, it is not like it is a lot of code

something like
https://github.com/SciTools/iris/blob/dask/lib/iris/fileformats/pp.py#L1040
```
    if mdi in data:
        if masked_array.dtype.kind == 'i':
            masked_array = masked_array.astype(np.dtype('f8'))
        data[data == mdi] = np.nan
```
perhaps with a code comment that this logic is reproduced here to avoid creating transient arrays or somesuch